File: lib/src/internal/routes/otp.dart
import '/src/sdk/routes.dart';
import '/src/types/others.dart';
import '/src/types/responses.dart';
import '../http/descope_client.dart';
import 'shared.dart';

class Otp implements DescopeOtp {
  final DescopeClient client;

  Otp(this.client);

  @override
  Future<String> signUp({required DeliveryMethod method, required String loginId, SignUpDetails? details}) async {
    return (await client.otpSignUp(method, loginId, details)).convert(method);
  }

  @override
  Future<String> signIn({required DeliveryMethod method, required String loginId, SignInOptions? options}) async {
    return (await client.otpSignIn(method, loginId, options)).convert(method);
  }

  @override
  Future<String> signUpOrIn({required DeliveryMethod method, required String loginId, SignInOptions? options}) async {
    return (await client.otpSignUpIn(method, loginId, options)).convert(method);
  }

  @override
  Future<AuthenticationResponse> verify({required DeliveryMethod method, required String loginId, required String code}) async {
    return (await client.otpVerify(method, loginId, code)).toAuthenticationResponse();
  }

  @override
  Future<String> updateEmail({required String email, required String loginId, required String refreshJwt, UpdateOptions? options}) async {
    return (await client.otpUpdateEmail(email, loginId, refreshJwt, options)).convert(DeliveryMethod.email);
  }

  @override
  Future<String> updatePhone({required String phone, required DeliveryMethod method, required String loginId, required String refreshJwt, UpdateOptions? options}) async {
    return (await client.otpUpdatePhone(phone, method, loginId, refreshJwt, options)).convert(method);
  }
}

File: ios/descope/Sources/descope/descope-swift-sdk/internal/routes/OTP.swift

final class OTP: DescopeOTP {
    let client: DescopeClient
    
    init(client: DescopeClient) {
        self.client = client
    }

    func signUp(with method: DeliveryMethod, loginId: String, details: SignUpDetails?) async throws(DescopeError) -> String {
        return try await client.otpSignUp(with: method, loginId: loginId, details: details).convert(method: method)
    }
    
    func signIn(with method: DeliveryMethod, loginId: String, options: [SignInOptions]) async throws(DescopeError) -> String {
        let (refreshJwt, loginOptions) = try options.convert()
        return try await client.otpSignIn(with: method, loginId: loginId, refreshJwt: refreshJwt, options: loginOptions).convert(method: method)
    }
    
    func signUpOrIn(with method: DeliveryMethod, loginId: String, options: [SignInOptions]) async throws(DescopeError) -> String {
        let (refreshJwt, loginOptions) = try options.convert()
        return try await client.otpSignUpIn(with: method, loginId: loginId, refreshJwt: refreshJwt, options: loginOptions).convert(method: method)
    }
    
    func verify(with method: DeliveryMethod, loginId: String, code: String) async throws(DescopeError) -> AuthenticationResponse {
        return try await client.otpVerify(with: method, loginId: loginId, code: code).convert()
    }
    
    func updateEmail(_ email: String, loginId: String, refreshJwt: String, options: UpdateOptions) async throws(DescopeError) -> String {
        return try await client.otpUpdateEmail(email, loginId: loginId, refreshJwt: refreshJwt, options: options).convert(method: .email)
    }
    
    func updatePhone(_ phone: String, with method: DeliveryMethod, loginId: String, refreshJwt: String, options: UpdateOptions) async throws(DescopeError) -> String {
        return try await client.otpUpdatePhone(phone, with: method, loginId: loginId, refreshJwt: refreshJwt, options: options).convert(method: method)
    }
}
